client/client.go

```diff
@@ -247,26 +214,170 @@ func (c *Client) update(latestRoot bool) (data.TargetFiles, error) {
+			if err := c.db.UnmarshalExpired(nPlusOneRootMetadata, nPlusOneRootMetadataSigned, "root", nPlusOne); err != nil {
+				// 5.3.6 Note that the expiration of the new (intermediate) root
+				// metadata file does not matter yet, because we will check for
+				// it in step 5.3.10.
+				if _, ok := err.(verify.ErrExpired); !ok {
+					return err
+				}
+			}
+		}
+
+		// 5.3.5 Following up, we check for a fast-forward attack: here, we check for that nPlusOneRootMetadataSigned.version >= nPlusOne.
+		if nPlusOneRootMetadataSigned.Version != nPlusOne {
+			return verify.ErrWrongVersion{
+				Given:    nPlusOneRootMetadataSigned.Version,
+				Expected: nPlusOne,
+			}
+		}
+
+		// 5.3.7 Set the trusted root metadata file to the new root metadata file.
+		c.rootVer = nPlusOneRootMetadataSigned.Version
+		// NOTE: following up on 5.3.1, we want to always have consistent snapshots on for the duration
+		// of root rotation. AFTER the rotation is over, we will set it to the value of the last root.
+		consistentSnapshot = nPlusOneRootMetadataSigned.ConsistentSnapshot
+		// 5.3.8 Persist root metadata. The client MUST write the file to non-volatile storage as FILENAME.EXT (e.g. root.json).
+		// NOTE: Internally, setMeta stores metadata in LevelDB in a persistent manner.
+		if err := c.local.SetMeta("root.json", nPlusOneRootMetadata); err != nil {
+			return err
+		}
+
+		for k := range startKeyIDs {
+			if !reflect.DeepEqual(startKeyIDs[k], getKeyIDs(k)) {
+				c.local.SetMeta(k, json.RawMessage{})
+			}
+		}
+
+		// 5.3.4.2 check that N+1 signed itself.
+		//This is different from the previous call because now the threashold and the keys are updated.
+		if err := c.getLocalRootMeta(); err != nil {
+			// 5.3.6 Note that the expiration of the new (intermediate) root
+			// metadata file does not matter yet, because we will check for
+			// it in step 5.3.10.
+			if _, ok := err.(verify.ErrExpired); !ok {
+				return err
+			}
+		}
+
+		// 5.3.9 Repeat steps 5.3.2 to 5.3.9
+	}
+	// 5.3.10 Check for a freeze attack.
+	// NOTE: this will check for any, including freeze, attack.
+	if err := c.getLocalRootMeta(); err != nil {
+		return err
+	}
+
+	// 5.3.11 If the timestamp and / or snapshot keys have been rotated,
+	// then delete the trusted timestamp and snapshot metadata files.
+	for k := range startKeyIDs {
+		if !reflect.DeepEqual(startKeyIDs[k], getKeyIDs(k)) {
+			c.local.SetMeta(k, json.RawMessage{})
+		}
+	}
+
+	// 5.3.12 Set whether consistent snapshots are used as per the trusted root metadata file.
+	c.consistentSnapshot = consistentSnapshot
+	return nil
+}
+
 func (c *Client) updateWithLatestRoot(m *data.SnapshotFileMeta) (data.TargetFiles, error) {
 	var rootJSON json.RawMessage
 	var err error
-	if m == nil {
-		rootJSON, err = c.downloadMetaUnsafe("root.json", defaultRootDownloadLimit)
+	if m == nil && c.rootVer != 0 {
+		if err = c.updateRoots(); err != nil {
+			if _, ok := err.(ErrEmptyTimestampOrSnapshot); !ok {
```

Comment:
am i missing where this error is ever used? if i am, can you please comment when/where `updateRoots()` may return this?

Reply:
The whole `updateWithLatestRoot` is now obsolete. Removing the method altogether.